### ISSUE TITLE
The assistant_only_loss field becomes ineffective when packing=True is enabled.

### ISSUE THREAD

**Issue by GitMonkey0** [huggingface/trl#3728]
During training, when packing=False is enabled, each piece of data will have a completion_mask constructed to mask the relevant fields of template, user, and system. However, when packing=True is set, the completion_mask no longer appears, causing issues with the packing training method. As a result, the trained model will continuously output <|im_start|>.

**Comment by LeonEricsson** [comment]
From what I can gather, `assistant_only_loss` uses an `assistant_mask` to mask tokens. [The assistant mask shouldn't be affected by packing](https://github.com/huggingface/trl/blob/640a9f39164ce3f9bbac7d325c1149ba023314e6/trl/trainer/sft_trainer.py#L186-L256). Could you provide a minimal example to replicate your observed behaviour?

**Comment by GitMonkey0** [comment]
@LeonEricsson 
Just like in the example at https://github.com/huggingface/trl/blob/main/trl/scripts/sft.py, debug it using the following command:
```

python trl/scripts/sft.py \
  --model_name_or_path Qwen/Qwen2-0.5B \
  --dataset_name trl-lib/ultrafeedback-gpt-3.5-turbo-helpfulness \
  --learning_rate 2.0e-5 \
  --num_train_epochs 1 \
  --per_device_train_batch_size 2 \
  --gradient_accumulation_steps 8 \
  --gradient_checkpointing \
  --eos_token '<|im_end|>' \
  --eval_strategy no \
  --eval_steps 100 \
  --output_dir Qwen2-0.5B-SFT \
  --bf16 True \
  --attn_implementation flash_attention_2 \
  --assistant_only_loss True \
  --packing True
```



When you look into the DataCollatorForLanguageModeling, you'll notice that when packing is not enabled, the data has a completion_mask field. However, when packing is enabled, this completion_mask field is absent from the data.

**Comment by LeonEricsson** [comment]
Okay, I see what you mean now.
